File: leetcode/3001-4000/3201-3300/3264-final-array-state-after-k-multiplication-operations-i/FinalArrayStateAfterKMultiplicationOperationsI.ts
function getFinalState(numbers: number[], k: number, multiplier: number): number[] {
  while (k-- > 0) {
    const { index, value } = numbers.reduce((min, currentValue, currentIndex) => {
      const { value } = min;

      if (value > currentValue) {
        return {
          index: currentIndex,
          value: currentValue
        };
      }

      return min;
    }, { index: 0, value: numbers[0] });

    numbers[index] = value * multiplier;
  }
  return numbers;
};

const numbers = [1, 2];
const k = 3;
const multiplier = 4;

console.log(getFinalState(numbers, k, multiplier))

File: leetcode/3001-4000/3201-3300/3261-3270/3264-final-array-state-after-k-multiplication-operations-i/FinalArrayStateAfterKMultiplicationOperationsI.java
import java.util.Arrays;
import java.util.PriorityQueue;

public class FinalArrayStateAfterKMultiplicationOperationsI {
  public static void main(String[] args) {
    Solution s = new Solution();

    System.out.println(Arrays.toString(s.getFinalState(new int[] { 1, 2 }, 3, 4)));
  }
}

class Solution {
  public int[] getFinalState(int[] numbers, int k, int multiplier) {
    PriorityQueue<Integer> queue = new PriorityQueue<>(
        (a, b) -> numbers[a] != numbers[b] ? numbers[a] - numbers[b] : a - b);

    for (int i = 0; i < numbers.length; i++) {
      queue.offer(i);
    }

    while (k-- > 0) {
      int index = queue.poll();

      numbers[index] *= multiplier;

      queue.offer(index);
    }

    return numbers;
  }
}
